### PR TITLE
Custom options

### PR DESCRIPTION
**Pull request by seansfkelley**
Pulled in @georgwiese's changes from #35 and my changes from #60. Responses to your comments on #35:

- **If `name` is set, the hidden `<input>`'s value gets set to `[Object object]`.** Indeed. This PR still does that, but I'm not clear on the value of the hidden input so I haven't fixed it. If it should reflect the display value, that's an easy change to make. What's the point of it?
- **The new methods are only set on Typeahead, but not Tokenizer as well.** I don't use Tokenizer so I haven't dug into the internals at all. Before I or anyone else does that I would like to get some feedback on this PR, though I would prefer not to block on that if that's okay.
- **Does `getDisplayString` need to return a string? Or can it return jsx/html? if the latter, a name like renderOption might be more appropriate.** Unfortunately yes, because it has to set the value of the input field which of course must be a string. You could work around this by dividing up the "render" step from the "display string" step, but I didn't want to introduce unnecessary complexity for a feature I'm not yet convinced it worth it.
- **Since the props will be common to both, having a mixin to help avoid redundancy might make things cleaner** I'm not sure I understand -- Tokenizer is just forwarding a bunch of props, so what would a mixin provide? In fact, I think it should probably not re-specify all of the props available to Typeahead and simply forward its own props (minus the ones that are only for it, like the token-add callback) with a splat.

This should be trivially backwards compatible as well. Sorry the commit history is a little messy; merging two in-flight PRs from different authors and adding more on top doesn't lend itself to a nice history. I hope that's alright.

### CODE REVIEW

**Comment by mull** [comment]
:+1:

**Comment by mull** [comment]
@fmoo thoughts on merging this in?

**Comment by fmoo** [comment]
> If it should reflect the display value, that's an easy change to make. What's the point of it?

The point of it is to be able to allow selected tokenizer values to be submitted in an html form without additional javascript.

An additional custom callback for extracting a useful value from the token object for submission (E.g., the id of the object or some unique string key outside of the display value).

> Tokenizer is just forwarding a bunch of props, so what would a mixin provide?
> I don't use Tokenizer so I haven't dug into the internals at all.

The idea is that the mixin would provide the same set of functionality for the tokenizer as well, but it's not clear whether avoiding a little copy paste is worth the added complexity.  Better to just copy-paste these callbacks to allow them to be user defined on the tokenizer as well.

**Comment by seansfkelley** [comment]
> The point of it is to be able to allow selected tokenizer values to be submitted in an html form without additional javascript.
> An additional custom callback for extracting a useful value from the token object for submission (E.g., the id of the object or some unique string key outside of the display value).

Ah right, form submission. The most flexible way is probably the way you mention but it's unfortunate this requires a whole new bit of the API to implement. :( Two questions: what should the default behavior be in the case where no such callback is provided (display string?) and does that imply that we can avoid using an extra option and just use the default behavior all the time? In other words, would it be bad to just always use the display string?

> ...it's not clear whether avoiding a little copy paste is worth the added complexity. Better to just copy-paste these callbacks to allow them to be user defined on the tokenizer as well.

I'll take a look at doing this.

File: src/typeahead/index.js [hunk 1]
```diff
@@ -238,6 +242,41 @@ var Typeahead = React.createClass({
...
+    var filterOptionProp = this.props.filterOption;
+    if (typeof filterOptionProp === 'function') {
+      return function(value, options) {
+        return options.filter(function(o) { return filterOptionProp(value, o); });
+      };
+    } else {
+      var mapper;
+      if (typeof filterOptionProp === 'string') {
+        mapper = function(o) { return o[filterOptionProp]; };
+      } else {
```
Comment (fmoo) [inline src/typeahead/index.js]: Can this ever happen?  I thought we just checked if typeof filterOptionProp was 'function' up above?

File: src/typeahead/index.js [hunk 2]
```diff
@@ -36,8 +43,7 @@ var Typeahead = React.createClass({
       defaultValue: "",
       placeholder: "",
       onOptionSelected: function(option) {},
-      onKeyDown: function(event) {},
-      filterOption: null
```
Comment (fmoo) [inline src/typeahead/index.js]: Why get rid of the defaults?

**Comment by fmoo** [comment]
> what should the default behavior be in the case where no such callback is provided (display string)?

Display string is a good default, yeah.

> does that imply that we can avoid using an extra option and just use the default behavior all the time? In other words, would it be bad to just always use the display string?

The common (default) case for basic typeaheads / tokenizers will be to use the default behavior, but there's still value in having a custom input value for the tokenizer.  For a friends list tokenizer, I'd want to be able to select a friend based on their name, but for the POST, I'd want the data to contain their user_id instead. 

> I'll take a look at doing this.

There's no need to follow up on the mixin approach if you don't want to.  I think mixins add a lot of extra complexity and the amount of copy-paste here is pretty minor.

Lastly, to get back to my earlier point about a `renderTypeaheadOption`; eventually, we'll want the typeahead *options* themselves (not necessarily the text entry) to be richer than just basic text (e.g., to support partial bolding of the matched text, or the profile pictures of the users you can select from).   We can add that later when someone needs it though.

**Comment by seansfkelley** [comment]
Added a string/function in the style of filterOption/displayOption called formInputOption.

Haven't gotten around to pushing these changes through to tokenizer, cause I don't use it. I'll take a quick look to make sure I didn't do anything dumb there.

Re: react components for display, I totally agree. In fact, with the addition of formInputOption that must return a string, you could potentially repurpose/overload it so it's the value used for al under-the-hood operations and displayOption can freely return a component. I don't use that behavior though, so I did not implement it (plus want to keep this PR as focused as possible).

File: src/typeahead/index.js (hunk 2 above)
Comment (seansfkelley) [inline src/typeahead/index.js]: It's a stylistic thing, mostly. I find specifying `null` or `undefined` as defaults noisy, but I can restore it if you prefer to have explicit defaults listed for all non-required options.

File: src/typeahead/index.js (hunk 1 above)
Comment (seansfkelley) [inline src/typeahead/index.js]: Yeah, this is the null case. I tweaked some names to make it clearer what the intent is here.

**Comment by jerryf196** [comment]
Is this released in the npm package? If so how can i use this?

**Comment by seansfkelley** [comment]
@jerryf196 since the compiled resources are only published during the build, I recommend forking, checking out this branch, running the build yourself and then having an npm dependency on that fork.

**Comment by fmoo** [comment]
Ok, I did a 1.0.8 release with this change in it.

**Comment by raysuelzer** [comment]
Thank you @seansfkelley !

**Pull request merged**